plugins/single_plugins/preserve-output.cpp

```diff
@@ -0,0 +1,314 @@
+    view->get_data_safe<last_output_info_t>()->geometry   = view->get_wm_geometry();
+    view->get_data_safe<last_output_info_t>()->fullscreen = view->fullscreen;
+    view->get_data_safe<last_output_info_t>()->minimized  = view->minimized;
+    view->get_data_safe<last_output_info_t>()->sticky     = view->sticky;
+    view->get_data_safe<last_output_info_t>()->tiled_edges = view->tiled_edges;
+    view->get_data_safe<last_output_info_t>()->z_order     = z_order;
+    if (view == output->get_active_view())
+    {
+        view->get_data_safe<last_output_info_t>()->focused = true;
+    }
+}
+
+nonstd::observer_ptr<last_output_info_t> view_get_data(wayfire_view view)
+{
+    return view->get_data<last_output_info_t>();
+}
+
+bool view_has_data(wayfire_view view)
+{
+    return view->has_data<last_output_info_t>();
+}
+
+void view_erase_data(wayfire_view view)
+{
+    view->erase_data<last_output_info_t>();
+}
+
+struct last_output_info_z_sort
+{
+    inline bool operator ()(nonstd::observer_ptr<wf::view_interface_t> view1,
+        nonstd::observer_ptr<wf::view_interface_t> view2)
+    {
+        return view_get_data(view1)->z_order > view_get_data(view2)->z_order;
+    }
+};
+
+/**
+ * Core last focused output
+ */
+
+class last_focused_output_t : public wf::custom_data_t
+{
+  public:
+    std::string output_identifier = "";
+    std::chrono::time_point<std::chrono::steady_clock> timestamp;
+};
+
+bool core_focused_output_expired()
+{
+    return std::chrono::duration_cast<std::chrono::milliseconds>(
+        std::chrono::steady_clock::now() -
+        wf::get_core().get_data_safe<last_focused_output_t>()->timestamp).count() >
+           10000;
+}
+
+void core_store_focused_output(wf::output_t *output)
+{
+    // Store the output as last focused if no other output has been stored as last
+    // focused in the last 10 seconds
+    if ((wf::get_core().get_data_safe<last_focused_output_t>()->output_identifier ==
+         "") || core_focused_output_expired())
+    {
+        LOGD("Setting last focused output to: ", output->to_string());
+        wf::get_core().get_data_safe<last_focused_output_t>()->output_identifier =
+            make_output_identifier(output);
+        wf::get_core().get_data_safe<last_focused_output_t>()->timestamp =
+            std::chrono::steady_clock::now();
+    }
+}
+
+std::string core_get_focused_output()
+{
+    return wf::get_core().get_data_safe<last_focused_output_t>()->output_identifier;
+}
+
+void core_erase_focused_output()
+{
+    wf::get_core().erase_data<last_focused_output_t>();
+}
+
+/**
+ * preserve-output plugin
+ */
+
+class wayfire_preserve_output : public wf::plugin_interface_t
+{
+    std::vector<std::string> outputs_being_removed;
+
+    wf::signal_connection_t output_pre_remove = [=] (wf::signal_data_t *data)
+    {
+        auto signal_data = (wf::output_pre_remove_signal*)data;
+        LOGD("Received pre-remove event: ", signal_data->output->to_string());
```

Comment:
I think core would emit the pre-remove event, destroy the output, emit the removed event? So why do you need to keep track of which outputs are being removed?

Reply:
This is because we're listening for move events on the views, to see if the user has explicitly moved them. I was keeping track of whether or not outputs are currently being destroyed because if they are, the move event is probably coming from the view being automatically moved. However this is actually left over from before I changed it to listen for `view_move_request_signal` rather than `view_geometry_changed_signal` and possibly the former is not emitted when the views are being automatically moved?

Reply:
I think you should indeed listen for the view_geometry_changed signal, after all, if the user invokes grid plugin for a view, there won't be any `view_move_request` signal but we don't want to restore to the old output, right?

But also my point was that you don't need to store a vector, a simple `bool output_is_being_removed` should suffice because it is always pre-remove, remove, pre-remove, remove

Reply:
Ok, so there won't ever be a situation where there are two outputs being removed at once? Makes sense.

The problem with listening for view_geometry_changed is that rarely an application will actually emit the signal after the output has been removed. Maybe we should think of some other means of invalidating the last_output_info on the views. If the user is sat there using their computer while the output is off that implies the output has been deliberately turned off rather than disconnected temporarily due to DPMS. Then again they could be moving their mouse and pressing keys to wake up the computer so we can't just listen for that.

Reply:
> Ok, so there won't ever be a situation where there are two outputs being removed at once? Makes sense.

They can be removed one after the other but not "concurrently" so to say.

Reply:
@ammen99 - I've thought about this and think we can make this simpler and more robust.

We keep track of the last time an output-pre-remove or an output-remove event was emitted. We have a callback on the view geometry changed which removes the view's last output info if an output [pre]-remove event wasn't emitted in the last X ms. Yes this involves another arbitrary timeout, but as I say I've seen Termite emit geometry changed events due to being moved to another output AFTER the output remove event was triggered, so I think this would be more robust.